analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    [RuleParameter("debugThreshold", PropertyType.Integer, "The maximum number of DEBUG, TRACE and VERBOSE statements allowed in the same code block.", DefaultTresholds.Debug)]
+    public int DebugThreshold { get; set; } = DefaultTresholds.Debug;
+
+    [RuleParameter("errorThreshold", PropertyType.Integer, "The maximum number of ERROR and FATAL statements allowed in the same code block.", DefaultTresholds.Error)]
+    public int ErrorThreshold { get; set; } = DefaultTresholds.Error;
+
+    [RuleParameter("informationThreshold", PropertyType.Integer, "The maximum number of INFORMATION statements allowed in the same code block.", DefaultTresholds.Information)]
+    public int InformationThreshold { get; set; } = DefaultTresholds.Information;
+
+    [RuleParameter("warningThreshold", PropertyType.Integer, "The maximum number of WARNING statements allowed in the same code block.", DefaultTresholds.Warning)]
+    public int WarningThreshold { get; set; } = DefaultTresholds.Warning;
+
+    protected override void Initialize(SonarParametrizedAnalysisContext context) =>
+        context.RegisterCompilationStartAction(cc =>
+        {
+            if (cc.Compilation.ReferencesAny(KnownAssembly.CastleCore, KnownAssembly.Log4Net, KnownAssembly.MicrosoftExtensionsLoggingAbstractions, KnownAssembly.Serilog, KnownAssembly.NLog))
+            {
+                cc.RegisterNodeAction(Process, SyntaxKind.Block, SyntaxKind.CompilationUnit);
+            }
+        });
+
+    private void Process(SonarSyntaxNodeReportingContext context)
+    {
+        var node = context.Node;
+        if (IsBlockNodeSupported(node))
+        {
+            var logCallCollector = new LoggingCallCollector(context.SemanticModel, node);
+            logCallCollector.Visit(node);
+            foreach (var loggingCategory in logCallCollector.LoggingInvocations)
+            {
+                var treshold = Treshold(loggingCategory.Key);
+                var invocations = loggingCategory.Value;
+                if (invocations.Count > treshold)
+                {
+                    var primaryLocation = invocations[0].GetLocation();
+                    var secondaryLocations = invocations.Skip(1).Select(x => x.GetLocation());
+                    context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, secondaryLocations, loggingCategory.Key, invocations.Count, treshold));
+                }
+            }
+        }
+    }
+
+    private int Treshold(string category)
+    {
+        var threshold = category switch
+        {
+            CategoryNames.Debug => DebugThreshold,
+            CategoryNames.Error => ErrorThreshold,
+            CategoryNames.Warning => WarningThreshold,
+            _ => InformationThreshold
+        };
+        return Math.Max(threshold, 0);
```

Comment:
This looks awfully like a dictionary.
Is there any reason you chose a lookup method instead?

If it's about lazy-initializing because of the user parameters, you could do it inside `CompilationStart` action.

Reply:
Yes, it's because of the parameterization. I'd rather keep it that way. If I put it to `CompilationStart` then the dictionary would be re-created for every project in the solution.

Reply:
sure, I don't mind that at all